File: sed_repos/java/SED3/src/main/java/ic/doc/strategy/Sequence.java
package ic.doc.strategy;

import java.util.Iterator;

public class Sequence implements Iterable<Integer> {

    private final TermGenerator termGenerator;

    public Sequence(TermGenerator termGenerator) {
        this.termGenerator = termGenerator;
    }

    public int term(int i) {
        if (i < 0) {
            throw new IllegalArgumentException("Not defined for indices < 0");
        }
        return termGenerator.term(i);
    }

    public Iterator<Integer> iterator() {
        return new SequenceIterator();
    }

    private class SequenceIterator implements Iterator<Integer> {

        private int index = 0;

        @Override
        public boolean hasNext() {
            return true;
        }

        @Override
        public Integer next() {
            return term(index++);
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException("Remove is not implemented");
        }
    }
}

File: sed_repos/python/SED3/src/strategy/sequence.py
from src.strategy.term_generator import TermGenerator
from typing import Iterator


class Sequence:
    def __init__(self, term_generator: TermGenerator):
        self.term_generator = term_generator

    def term(self, i: int) -> int:
        if i < 0:
            raise ValueError("Not defined for indices < 0")
        return self.term_generator.term(i)

    def __iter__(self) -> Iterator[int]:
        return self.SequenceIterator(self)

    class SequenceIterator:
        def __init__(self, sequence):
            self.sequence = sequence
            self.index = 0

        def __iter__(self):
            return self

        def __next__(self) -> int:
            result = self.sequence.term(self.index)
            self.index += 1
            return result
